keps/sig-apps/2214-indexed-job/README.md

```diff
@@ -0,0 +1,487 @@
+
+The lack of support for this pattern in k8s forces users to implement their
+own APIs and controllers or adopt third party implementations. Each
+implementation splits the ecosystem, making it harder for higher level systems
+for Job queueing or workflows to support all of them.
+
+[Job patterns]: https://kubernetes.io/docs/concepts/workloads/controllers/job/#job-patterns
+
+### Goals
+
+- Support the *array Job* pattern by adding completion indexes to each Pod
+  of a Job in *fixed completion count* mode.
+
+### Non-Goals
+
+- Support for work lists, where each Pod receives a different element of a
+  static list. This can be implemented by users from completion indexes.
+- Support for completion index in non-parallel Jobs or Jobs with a work queue.
+- All-or-nothing scheduling.
+
+## Proposal
+
+### User Stories (Optional)
+
+#### Story 1
+
+As a Job author, I can create an array Job where each Pod receives a unique and
+ordered completion index. I can use the index in my binary through an
+environment variable or a file.
+
+```yaml
+apiVersion: batch/v1
+kind: Job
+metadata:
+  name: parallel-work
+spec:
+  completions: 100
+  parallelism: 100
+  template:
+    spec:
+      containers:
+      - name: task
+        image: registry.example.com/processing-image
+        command: ["./process",  "--index", "$INDEX"]
+        env:
+        - name: INDEX
+          valueFrom:
+            fieldRef:
+              fieldPath: metadata.annotations['alpha.batch.k8s.io/job-completion-index'] 
+```
+
+### Notes/Constraints/Caveats (Optional)
+
+* An earlier proposal for [indexed Job]
+suggested the support for work lists, i.e. passing different parameters to each
+Pod. We decided to leave this out of the proposal to keep it simple and
+because work lists can be implemented in a startup script using the completion
+index as building block.
+* The semantics of an indexed Job are similar to a StatefulSet, in the sense
+that Pods have an associated index and Pods.
+However, the APIs have a major difference: a StatefulSet doesn't have completion
+semantics, as opposed to Jobs.
+
+[indexed Job]: https://github.com/kubernetes/community/blob/b21d1b27c8c748bf81283c2d89cde2becb5f2709/contributors/design-proposals/apps/indexed-job.md
+
+### Risks and Mitigations
+
+Jobs have a known issue in which more than one Pod can be started even if
+parallelism and completion are set to 1 ([reference]). In the case of indexed
+Jobs, this translates as more than one Pod having the same index.
+
+Just like for existing Job patterns, workloads have to handle duplicates at the
+application level.
+
+[reference]: https://kubernetes.io/docs/concepts/workloads/controllers/job/#handling-pod-and-container-failures
+
+## Design Details
+
+### Indexed Jobs
+
+Indexed Jobs are Jobs [with fixed completion count], that is, Jobs that have a
+non-nil `.spec.completions` field.
+
+### Pod detail
+
+The Pod and PodSpec APIs don't get any new fields. However, Pods created for
```

Comment:
Did you consider the alternative where you do change the API such as by adding `.spec.indexed` boolean?

Some pros of that approach are:

1. Could use a different path in the control code for indexed jobs.  Reduces risk that the new path in the controller hurts existing multi-completition jobs that don't need indexes.
1. Upgrade downgrade is simple:  On downgrade, all the ones with ".spec.indexed" stop working.  On upgrade, those without ".spec.indexed" don't do anything different.
2. Can inject index without modifying existing pods see subsequent comments).

Reply:
@erictune 

1) - I think that with the field added to job explicitly you can also separate code path cleanly. We have the necessary information in the job, and for pods we can check by OwnerRef.

2) Eric - can you clarify the part of "On downgrade, all the ones with ".spec.indexed" stop working."?
  I guess I'm missing the point what exactly this boolean will be responsible for controlling?
  Setting the env var of the index? Something more?

On a related note - if modifying PodSpec, when not add "index" int field, but just bool indexed one?

Reply:
Answering in upgrade/downgrade section.

Reply:
+1, it can also be used to create a metric to track usage.

Reply:
My argument against having a `.spec.indexed` boolean field is because of the possible extension to KEP: "stable pod names" or "Completion Index in the Pod Name" (listed in Alternatives section).

In that case, API validation becomes more tricky, which in turns makes the API more unusable:
- `.spec.indexed=true` requires `.spec.completions!=nil` (with current KEP)
- `.spec.stableNames=true` requires `.spec.indexed=true` (extension to the KEP)

But, maybe we can do the following (kind of back to the first version of this KEP):
Add the field `.spec.completionMode` which can have values: NonIndexed (default), Indexed, IndexedAndUnique.

Reply:
Added back `spec.completionMode` to the proposal. PTAL

Reply:
That could work.  Another option is as follows:

1. `.spec.indexed` and  `.spec.stableNames` both unset or false :arrow_right: Non-indexed behavior.
2. `.spec.indexed` true and  `.spec.stableNames` set unset   :arrow_right: Indexed behavior
3. `.spec.indexed`  and  `.spec.stableNames` both set true :arrow_right:  IndexedAndUnique behavior.
4. `.spec.indexed` unset and  `.spec.stableNames` set true :arrow_right: admission defaults `.spec.indexed` to true for you and you get  IndexedAndUnique behavior.
5. Other combinations are an error.

Whether the user wants Indexed, or IndexedAndUnique, they only have to type one line of yaml.

Reply:
If it were me, I would not attempt to validate or default these fields in OpenAPI.  I would do this defaulting and validation it in admission or the controller.

Reply:
Once you add an enum to an API, you cannot add new values to it.
See https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/api_changes.md#backward-compatibility-gotchas
But you can add new fields.  Which is why I suggest the new-field-per-feature field.

Reply:
Oh API nuances :(

But if I protect the new value with a feature gate, I don't see it much different from adding a new field. @liggitt ?

Reply:
> @erictune:
> Once you add an enum to an API, you cannot add new values to it. See https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/api_changes.md#backward-compatibility-gotchas

That also gives a path for enum fields we expect to expand in the future:

> For enumeration-like fields that expect to add new values in the future, such as reason fields, please document that expectation clearly in the API field descriptions. Clients should treat such sets of values as potentially open-ended.

We encourage people to use enums over booleans for future expansion purposes in https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/api-conventions.md#primitive-types:

> Think twice about bool fields. Many ideas start as boolean but eventually trend towards a small set of mutually exclusive options. Plan for future expansions by describing the policy options explicitly as a string type alias (e.g. TerminationMessagePolicy).

Reply:
Going with enum. See updated API documentation.